src/assert/assert_subroutine_m.F90

```diff
@@ -1,33 +1,48 @@
+!     (c) 2024 UC Regents, see LICENSE file for detailed terms.
 !
 !     (c) 2019-2020 Guide Star Engineering, LLC
 !     This Software was developed for the US Nuclear Regulatory Commission (US NRC) under contract
 !     "Multi-Dimensional Physics Implementation into Fuel Analysis under Steady-state and Transients (FAST)",
 !     contract # NRC-HQ-60-17-C-0007
 !
+#include "assert_macros.h"
+
 module assert_subroutine_m
-  !! summary: Utility for runtime checking of logical assertions.
+  !! summary: Utility for runtime enforcement of logical assertions.
   !! usage: error-terminate if the assertion fails:
   !!
   !!    use assertions_m, only : assert
   !!    call assert( 2 > 1, "2 > 1")
   !!
-  !! Turn off assertions in production code by setting USE_ASSERTIONS to .false. via the preprocessor.
+  !! Assertion enforcement is controlled via the `ASSERTIONS` preprocessor macro,
+  !! which can be defined to non-zero or zero at compilation time to
+  !! respectively enable or disable runtime assertion enforcement.
+  !!
+  !! When the `ASSERTIONS` preprocessor macro is not defined to any value,
+  !! the default is that assertions are *disabled* and will not check the condition.
+  !!
+  !! Disabling assertion enforcement may eliminate any associated runtime
+  !! overhead by enabling optimizing compilers to ignore the assertion procedure
+  !! body during a dead-code-removal phase of optimization.
+  !!
+  !! To enable assertion enforcement (e.g., for a debug build), define the preprocessor ASSERTIONS to non-zero.
   !! This file's capitalized .F90 extension causes most Fortran compilers to preprocess this file so
-  !! that building as follows turns off assertion enforcement: 
+  !! that building as follows enables assertion enforcement: 
   !!
-  !!    fpm build --flag "-DUSE_ASSERTIONS=.false."
+  !!    fpm build --flag "-DASSERTIONS"
   !!
-  !! Doing so may eliminate any associated runtime overhead by enabling optimizing compilers to ignore
-  !! the assertion procedure body during a dead-code-removal phase of optimization.
   implicit none
   private
   public :: assert
 
 #ifndef USE_ASSERTIONS
-# define USE_ASSERTIONS .true.
+#  if ASSERTIONS
+#    define USE_ASSERTIONS .true.
```

Comment:
I think block of code (lines 38-43) exists to support a pattern used in the Berkeley Lab fork of ICAR.  I would be fine with changing ICAR to our new recommended way of invoking assertions so that we don't need this. @bonachea   I think we can therefore drop this block if you agree with that.

Reply:
Alternatively, I could just copy this block of code into ICAR.

Reply:
As discussed verbally, this code just converts the `ASSERTIONS` knob into the form needed this module. I kept the same `USE_ASSERTIONS` name to avoid needlessly breaking old clients, but we no longer document that knob and new code should be encouraged to use the newly documented `ASSERTIONS` knob.